presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeyValuePairs.java

```diff
@@ -103,4 +118,63 @@ public void add(Block key, Block value, int keyPosition, int valuePosition)
             }
         }
     }
+
+    private boolean keyExists(Block key, int position)
+    {
+        checkArgument(position >= 0, "position is negative");
+        return keyPositionByHash[getHashPositionOfKey(key, position)] != EMPTY_SLOT;
+    }
+
+    private void addKey(Block key, int position)
+    {
+        checkArgument(position >= 0, "position is negative");
+        keyType.appendTo(key, position, keyBlockBuilder);
+        int hashPosition = getHashPositionOfKey(key, position);
+        if (keyPositionByHash[hashPosition] == EMPTY_SLOT) {
+            keyPositionByHash[hashPosition] = keyBlockBuilder.getPositionCount() - 1;
+            if (keyBlockBuilder.getPositionCount() >= maxFill) {
+                rehash(maxFill * 2);
```

Comment:
It should be double the original size instead of `maxFill * 2`. Please refer to `BigintGroupByHash`.

Please fix `TypedHistogram` as well.